In [3]:
import pandas as pd                                                               #Importing pandas using the path link
ds_salaries = pd.read_csv(r"C:\Users\user\Desktop\ALTSCHOOL AFRICA\EXAM PROJECT\ds_salaries_dirty.csv")

In [4]:
type (ds_salaries)                  #Identifying the type of dataset

pandas.core.frame.DataFrame

# Task 1: Clean and Validate the Salary Data


In [6]:
clean_salaries = ds_salaries.copy()                  #Creating a copy to preserve original df

In [7]:
clean_salaries                          #Loading the dataset

work_year experience_level employment_type       job_title  salary_in_usd  \
0        2022               SE              CT  Data Scientist        42183.0   
1        2023               MI              FL  Data Scientist       190371.0   
2        2020               MI              CT              DS       173946.0   
3        2022              NaN              FT     ML Engineer       146336.0   
4        2022               MI              FL     ML Engineer      9999999.0   
..        ...              ...             ...             ...            ...   
95       2021              NaN              FL   AI Specialist       246386.0   
96       2021               MI              PT   AI Specialist        56641.0   
97       2023               EX              CT     ML Engineer       195656.0   
98       2021               EN              FT   AI Specialist       193817.0   
99       2020               SE              FL  Data Scientist       184165.0   

    remote_ratio company_location company_size  
0             50               IN            L  
1            100               IN            S  
2             50               IN            L  
3             50               DE            S  
4            100               NG            L  
..           ...              ...          ...  
95            50               GB            M  
96             0               US            M  
97             0              NaN            M  
98             0               NG            S  
99             0               US            L  

[100 rows x 8 columns]

In [8]:
clean_salaries.shape    #Identifying the shape of the dataset

(100, 8)

In [9]:
clean_salaries.info                                   #Identifying the information of the dataset

<bound method DataFrame.info of     work_year experience_level employment_type       job_title  salary_in_usd  \
0        2022               SE              CT  Data Scientist        42183.0   
1        2023               MI              FL  Data Scientist       190371.0   
2        2020               MI              CT              DS       173946.0   
3        2022              NaN              FT     ML Engineer       146336.0   
4        2022               MI              FL     ML Engineer      9999999.0   
..        ...              ...             ...             ...            ...   
95       2021              NaN              FL   AI Specialist       246386.0   
96       2021               MI              PT   AI Specialist        56641.0   
97       2023               EX              CT     ML Engineer       195656.0   
98       2021               EN              FT   AI Specialist       193817.0   
99       2020               SE              FL  Data Scientist       184165.0

In [10]:
clean_salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   work_year         100 non-null    int64  
 1   experience_level  90 non-null     object 
 2   employment_type   92 non-null     object 
 3   job_title         100 non-null    object 
 4   salary_in_usd     96 non-null     float64
 5   remote_ratio      100 non-null    int64  
 6   company_location  94 non-null     object 
 7   company_size      100 non-null    object 
dtypes: float64(1), int64(2), object(5)
memory usage: 6.4+ KB


In [11]:
clean_salaries.head()                 # checKing to see the first 5 rows

work_year experience_level employment_type       job_title  salary_in_usd  \
0       2022               SE              CT  Data Scientist        42183.0   
1       2023               MI              FL  Data Scientist       190371.0   
2       2020               MI              CT              DS       173946.0   
3       2022              NaN              FT     ML Engineer       146336.0   
4       2022               MI              FL     ML Engineer      9999999.0   

   remote_ratio company_location company_size  
0            50               IN            L  
1           100               IN            S  
2            50               IN            L  
3            50               DE            S  
4           100               NG            L

In [12]:
# Take a closer look at the work_year column
clean_salaries['work_year'].head(100)

0     2022
1     2023
2     2020
3     2022
4     2022
      ... 
95    2021
96    2021
97    2023
98    2021
99    2020
Name: work_year, Length: 100, dtype: int64

In [13]:
clean_salaries['work_year'].unique()                 #To check the unique numbers inside year

array([2022, 2023, 2020, 2021], dtype=int64)

In [14]:
clean_salaries.isnull().all()                    #using isnull to check for empty rows

work_year           False
experience_level    False
employment_type     False
job_title           False
salary_in_usd       False
remote_ratio        False
company_location    False
company_size        False
dtype: bool

In [15]:
clean_salaries.isnull().sum()                                  #checking for the sum of empty rows

work_year            0
experience_level    10
employment_type      8
job_title            0
salary_in_usd        4
remote_ratio         0
company_location     6
company_size         0
dtype: int64

In [16]:
# Rule: work_year - Integer. Year the salary was reported. Must not be missing.
clean_salaries.dropna(subset=['work_year'], inplace=True)
clean_salaries['work_year'] = clean_salaries['work_year'].astype(int)

In [17]:
clean_salaries['work_year'].head(100)

0     2022
1     2023
2     2020
3     2022
4     2022
      ... 
95    2021
96    2021
97    2023
98    2021
99    2020
Name: work_year, Length: 100, dtype: int32

In [18]:
# Take a closer look at the experience_level column
clean_salaries['experience_level'].head(100)

0      SE
1      MI
2      MI
3     NaN
4      MI
     ... 
95    NaN
96     MI
97     EX
98     EN
99     SE
Name: experience_level, Length: 100, dtype: object

In [19]:
### Cleaning experience_level
##Replacing missing or invalid values with 'EN' as it's the assumed entry-level default.
# Rule: experience_level - Categorical. One of 'EN', 'MI', 'SE', 'EX'. Replace missing/invalid with 'EN'.
valid_experience_levels = ['EN', 'MI', 'SE', 'EX']
def clean_experience_level(level):
    if level not in valid_experience_levels:
        return 'EN'
    return level
clean_salaries['experience_level'] = clean_salaries['experience_level'].apply(clean_experience_level)

In [20]:
clean_salaries['experience_level'].head(100)

0     SE
1     MI
2     MI
3     EN
4     MI
      ..
95    EN
96    MI
97    EX
98    EN
99    SE
Name: experience_level, Length: 100, dtype: object

In [21]:
# Take a closer look at the employment_type column
clean_salaries['employment_type'].head(100)

0     CT
1     FL
2     CT
3     FT
4     FL
      ..
95    FL
96    PT
97    CT
98    FT
99    FL
Name: employment_type, Length: 100, dtype: object

In [22]:
### Cleaning `employment_type`
##Replacing missing or invalid values with 'FT' as it's the assumed entry-level default.
# Rule: employment_type - Categorical. One of 'FT', 'PT', 'CT', 'FL'. Replace missing values with 'FT'.

valid_employment_types = ['FT', 'PT', 'CT', 'FL']
def clean_employment_type(emp_type):
    if emp_type not in valid_employment_types:
        return 'FT'
    return emp_type
clean_salaries['employment_type'] = clean_salaries['employment_type'].apply(clean_employment_type)

# Corrected line to address FutureWarning
clean_salaries['employment_type'] = clean_salaries['employment_type'].fillna('FT')

In [23]:
clean_salaries['employment_type'].head(100)

0     CT
1     FL
2     CT
3     FT
4     FL
      ..
95    FL
96    PT
97    CT
98    FT
99    FL
Name: employment_type, Length: 100, dtype: object

In [24]:
# Take a closer look at the job_title column
clean_salaries['job_title'].head(100)

0     Data Scientist
1     Data Scientist
2                 DS
3        ML Engineer
4        ML Engineer
           ...      
95     AI Specialist
96     AI Specialist
97       ML Engineer
98     AI Specialist
99    Data Scientist
Name: job_title, Length: 100, dtype: object

In [25]:
# Remove job titles with less than 3 characters 
def is_valid_title(title):
    if isinstance(title, str) and len(title.strip()) >= 3:
        return True
    return False

clean_salaries = clean_salaries[clean_salaries['job_title'].apply(is_valid_title)]

In [26]:
clean_salaries['job_title'].head(100)

0     Data Scientist
1     Data Scientist
3        ML Engineer
4        ML Engineer
5     Data Scientist
           ...      
95     AI Specialist
96     AI Specialist
97       ML Engineer
98     AI Specialist
99    Data Scientist
Name: job_title, Length: 95, dtype: object

In [86]:
# Take a closer look at the salary_in_usd column
clean_salaries['salary_in_usd'].head(100)

0      42183.0
1     190371.0
3     146336.0
4     154019.0
5      35539.0
        ...   
95    246386.0
96     56641.0
97    195656.0
98    193817.0
99    184165.0
Name: salary_in_usd, Length: 95, dtype: float64

In [30]:

### Cleaning salary_in_usd
# Replacing the extreme values (less than $10,000 or greater than $1,000,000) with the median salary to prevent distortion in analysis. 
# Missing values were also filled this way.

clean_salaries = clean_salaries.copy()
median_salary = clean_salaries['salary_in_usd'].median()

def clean_salary_in_usd(salary):
    if salary < 10000 or salary > 1000000:
        return median_salary
    return salary
clean_salaries['salary_in_usd'] = clean_salaries['salary_in_usd'].apply(clean_salary_in_usd)


In [31]:
clean_salaries['salary_in_usd'].head(100)

0      42183.0
1     190371.0
3     146336.0
4     154019.0
5      35539.0
        ...   
95    246386.0
96     56641.0
97    195656.0
98    193817.0
99    184165.0
Name: salary_in_usd, Length: 95, dtype: float64

In [32]:
# Take a closer look at the company_location column
clean_salaries['company_location'].head(100)

0      IN
1      IN
3      DE
4      NG
5      GB
     ... 
95     GB
96     US
97    NaN
98     NG
99     US
Name: company_location, Length: 95, dtype: object

In [33]:
# Cleaning the company_location 
clean_salaries['company_location'] = clean_salaries['company_location'].fillna('US')

In [34]:
clean_salaries['company_location'].head(100)              

0     IN
1     IN
3     DE
4     NG
5     GB
      ..
95    GB
96    US
97    US
98    NG
99    US
Name: company_location, Length: 95, dtype: object

In [35]:
# Take a closer look at the company_size column
clean_salaries['company_size'].head(100)

0     L
1     S
3     S
4     L
5     S
     ..
95    M
96    M
97    M
98    S
99    L
Name: company_size, Length: 95, dtype: object

In [36]:
# Cleaning the company_size
clean_salaries['company_size'] = clean_salaries['company_size'].fillna('M')

In [37]:
clean_salaries['company_size'].head(100)

0     L
1     S
3     S
4     L
5     S
     ..
95    M
96    M
97    M
98    S
99    L
Name: company_size, Length: 95, dtype: object

In [38]:
clean_salaries.info()                  # Final result

<class 'pandas.core.frame.DataFrame'>
Index: 95 entries, 0 to 99
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   work_year         95 non-null     int32  
 1   experience_level  95 non-null     object 
 2   employment_type   95 non-null     object 
 3   job_title         95 non-null     object 
 4   salary_in_usd     91 non-null     float64
 5   remote_ratio      95 non-null     int64  
 6   company_location  95 non-null     object 
 7   company_size      95 non-null     object 
dtypes: float64(1), int32(1), int64(1), object(5)
memory usage: 6.3+ KB


# Task 2: Salary Trends by Experience


In [40]:
# Group by experience level and employment type

salary_trends = clean_salaries.groupby(['experience_level', 'employment_type'], observed=True)['salary_in_usd'].mean().reset_index()

In [41]:
salary_trends

experience_level employment_type  salary_in_usd
0                EN              CT   77025.250000
1                EN              FL  160093.333333
2                EN              FT  132732.444444
3                EN              PT  166740.000000
4                EX              CT  166817.200000
5                EX              FL  163361.250000
6                EX              FT  175002.625000
7                EX              PT   68467.000000
8                MI              CT  170696.750000
9                MI              FL  128410.250000
10               MI              FT  116323.750000
11               MI              PT  131637.750000
12               SE              CT  127803.833333
13               SE              FL  129452.750000
14               SE              FT  139054.400000
15               SE              PT  170817.666667

Observations:

1a. Executive (EX) Experience Level: Generally shows the highest average salaries across most employment types, especially for Full-Time (FT) roles. 
b. Contract (CT) and Freelance (FL) also show high averages. 
c. Part-Time (PT) roles for EX level appear to be an outlier with a lower average.

2. Entry-Level (EN) Experience Level: Some entry-level roles (Freelance and Part-time) shows relatively high average salaries, sometimes even surpassing certain mid-level and senior-level categories. This could indicate specific in-demand skills even at an entry-level or niche roles.
    
3. Mid-Level (MI) and Senior-Level (SE) Experience Levels: These levels typically show an increase in average salaries compared to entry-level, particularly for Full-Time (FT) and Part-Time (PT) positions.
    
4. Employment Type Impact: Full-time (FT) roles generally have competitive average salaries across all experience levels. Freelance (FL) and Contract (CT) roles can also offer high compensation, especially for higher experience levels.

# Task 3: Remote Work Analysis


In [45]:
# Filter remote and on-site roles only
remote_roles = clean_salaries[clean_salaries['remote_ratio'] == 100]
onsite_roles = clean_salaries[clean_salaries['remote_ratio'] == 0]

# Compute average salary for each
remote_avg = remote_roles['salary_in_usd'].mean()
onsite_avg = onsite_roles['salary_in_usd'].mean()

# Create final comparison DataFrame
remote_salary_comparison = pd.DataFrame({
    'Work Mode': ['Remote', 'On-site'],
    'Average Salary (USD)': [round(remote_avg, 2), round(onsite_avg, 2)]
})

remote_salary_comparison


Work Mode  Average Salary (USD)
0    Remote             144734.37
1   On-site             153214.04

Observations:

Based on this analysis, on-site roles appear to have a slightly higher average salary ($153,214.04) compared to fully remote roles ($144,734.37).

This difference, while not extremely large, suggests that, on average, companies might offer slightly higher compensation for roles that require physical presence in an office. This could be due to various factors, including:

1. Location-based compensation: On-site roles might be situated in higher cost-of-living areas, which typically correlate with higher salaries.

2. Role characteristics: Certain types of roles that naturally command higher salaries might also be more likely to be on-site due to specific collaboration needs, access to specialized equipment, or security requirements.

3. Company policies and benefits: Some companies that prioritize on-site work might offer better overall compensation packages, or certain benefits that are tied to office presence.

It's important to note that this is an average across all job titles and experience levels within the dataset. A more detailed breakdown (e.g., by specific job title or experience level for remote vs. on-site) could provide deeper insights into these trends.
